Python replacement for DTC [DNM/WIP]

nashif opened the pull request:
just for sharing here, still need to resolve some of the issues with a few boards and fix the last commit to be less hackish

lpereira committed:
scripts: dts: devicetree: Accept a list with include paths in parse_file()

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
scripts: dts: devicetree: Recursively update nodes when parsing

When parsing pre-processed .dts files, some nodes appear twice in the
file, causing some nodes to not appear in the parsed data.  Update the
nodes recursively.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
scripts: dts: devicetree: Remove 'type' from node

It was not being used by any script, and parse_node() was setting it
to the `type` variable, which contains the type `type` from Python.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
scripts: dts: Add a script to flatten a .dts file

This script parses a .dts file preprocessed by a C preprocessor and
dumps it as a .dts.  It's simlar to what the dtc compiler does, but
without requiring a host dtc binary (which might not be available on
all platforms Zephyr supports).

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
scripts: dts: devicetree: Ignore C preprocessor directives

DTS files may be preprocessed by the C preprocessor in order to build
nodes selectively based on configuration.  When CPP touches a file,
it leaves some directives that may be used by a compiler for errors
and such; just ignore these lines as we're not going to need them.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
scripts: dts: devicetree: Parse nodes declared in the same line

Some files contain lines such as "chosen { };", which were not being
correctly parsed.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
scripts: dts: devicetree: Empty properties should be marked as such

DTS syntax allows for the following:

	some-node {
		property-name;
	}

And the devicetree.py was converting property-name to a boolean
property with value 1.  In order to dump the DTS file, store
`{empty: True}` instead.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

SebastianBoe committed:
cmake: Use flatten.py instead of DTC

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: devicetree.py: Evaluate expressions

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: extract_dts_includes: Don't assume my phandle

With 96b_nitrogen on hello world I was observing a key error. This
patch resolves the key error.

With DTC interrupt-parent would always be set to a phandle, but with
flatten.py it is set to the name of the peripheral instead.

e.g.

        interrupt-parent = "/soc/interrupt-controller@e000e100";

vs

        interrupt-parent = <0x1>;

I don't know how all this works, so they may very well be a cleaner
solution here.

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: devicetree: Support more operators

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: devicetree: Support multiple expressions on the same line

Signed-off-by: Sebastian Boe <sebastian.boe@nordicsemi.no>

nashif committed:
hacks: apply very crude hacks to match dtc

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented:
Please add:

```
if (NOT WIN32)
endif()
```

around the `find_program` call in `cmake/host-tools.cmake`

carlescufi committed:
cmake: Don't look for dtc on Windows

With DTC now being handled with Python, there's no need to look for it
on Windows at all. Later we might even remove the requirement on UNIX
platforms.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
This is not required for v1.11 since we have all pieces we need now to build natively on Windows, including `dtc`.

nashif commented:
not needed anymore
